src/terminal/adapter/SixelParser.hpp

```diff
@@ -0,0 +1,125 @@
+        static size_t MaxColorsForLevel(const VTInt conformanceLevel) noexcept;
+
+        SixelParser(AdaptDispatch& dispatcher, const StateMachine& stateMachine, const VTInt conformanceLevel = DefaultConformance) noexcept;
+        void SoftReset();
+        void SetDisplayMode(const bool enabled) noexcept;
+        std::function<bool(wchar_t)> DefineImage(const VTInt macroParameter, const DispatchTypes::SixelBackground backgroundSelect, const VTParameter backgroundColor);
+
+    private:
+        // NB: If we want to support more than 256 colors, we'll also need to
+        // change the IndexType to uint16_t, and use a bit field in IndexPixel
+        // to retain the 16-bit size.
+        static constexpr size_t MAX_COLORS = 256;
+        using IndexType = uint8_t;
+        struct IndexedPixel
+        {
+            uint8_t transparent = false;
+            IndexType colorIndex = 0;
+        };
+
+        AdaptDispatch& _dispatcher;
+        const StateMachine& _stateMachine;
+        const VTInt _conformanceLevel;
+
+        void _parseCommandChar(const wchar_t ch);
+        void _parseParameterChar(const wchar_t ch);
+        int _applyPendingCommand();
+        void _executeCarriageReturn() noexcept;
+        void _executeNextLine();
+        void _executeMoveToHome();
+
+        enum class States : size_t
+        {
+            Normal,
+            Attributes,
+            Color,
+            Repeat
+        };
+        States _state = States::Normal;
+        std::vector<VTParameter> _parameters;
+
+        bool _initTextBufferBoundaries();
+        void _initRasterAttributes(const VTInt macroParameter, const DispatchTypes::SixelBackground backgroundSelect) noexcept;
+        void _updateRasterAttributes(const VTParameters& rasterAttributes);
+        void _scrollTextBuffer(Page& page, const int scrollAmount);
+        void _updateTextCursor(Cursor& cursor) noexcept;
+
+        const til::size _cellSize;
+        bool _displayMode = true;
+        til::rect _textMargins;
+        til::point _textCursor;
+        bool _textCursorWasVisible;
+        til::CoordType _availablePixelWidth;
+        til::CoordType _availablePixelHeight;
+        til::CoordType _maxPixelAspectRatio;
+        til::CoordType _pixelAspectRatio;
+        til::CoordType _sixelHeight;
+        til::CoordType _segmentHeight;
+        til::CoordType _pendingTextScrollCount;
+        til::size _backgroundSize;
+        bool _backgroundFillRequired;
+
+        void _initColorMap(const VTParameter backgroundColor);
+        void _defineColor(const VTParameters& colorParameters);
+        void _defineColor(const size_t colorNumber, const COLORREF color);
+        COLORREF _colorFromIndex(const IndexType tableIndex) const noexcept;
+        static constexpr RGBQUAD _makeRGBQUAD(const COLORREF color) noexcept;
+        void _updateTextColors();
+
+        std::array<size_t, MAX_COLORS> _colorMap = {};
+        std::array<bool, MAX_COLORS> _colorMapUsed = {};

```

Comment:
It seems as if the `_colorMap` items are supposed to be `IndexType`s. Should we swap `size_t` with `IndexType` here?

Side note: As an alternative you could store the `_colorMap` items as `int` and initialize them to a negative value to indicate that they're currently not used.

Reply:
Good point. I've change it to `IndexType` now. We do need to maintain the used state separately, though, because the color map is still preinitialized with default values when it's unused. It's only when those defaults are overridden that they become "used".